pkg/kubelet/deviceplugin/manager.go

```diff
@@ -212,21 +213,38 @@ func (m *ManagerImpl) addEndpoint(r *pluginapi.RegisterRequest) {
 		return
 	}
 
-	stream, err := e.list()
-	if err != nil {
-		glog.Errorf("Failed to List devices for plugin %v: %v", r.ResourceName, err)
-		return
-	}
+	closeChan := make(chan bool)
+	go e.run(closeChan)
 
 	go func() {
-		e.listAndWatch(stream)
-
-		m.mutex.Lock()
-		if old, ok := m.Endpoints[r.ResourceName]; ok && old == e {
-			delete(m.Endpoints, r.ResourceName)
+		// TODO: Consider making timeout value configurable.
+		t := time.NewTimer(5 * time.Second)
+		defer t.Stop()
+
+		for {
+			select {
+			case <-t.C:
+				m.mutex.Lock()
+				if old, ok := m.Endpoints[r.ResourceName]; ok && old == e {
```

Comment:
I usually use the pattern:

```
if !condition1 {
     return
}
if !condition2 {
     return
}
action()
}
```

Instead of the pattern:
```
if condition1 {
    if condition2 {
         action()
    }
}
if condition2 {
     return
}
action
}
```

It makes the code less indented which feels easier to read and makes it easier later to add `if !condition3`